### PR TITLE
changes to make zenohpico compatible with zenoh-cpp

### PR DESCRIPTION
**Pull request by milyin**
- missing generics and templates added, macros file made similar to zenohc one
- few missing API functions and structures added (timestamp, closure call functions)
- consts added, some fields renamed and changed places to make initalizer works universally

### CODE REVIEW

File: include/zenoh-pico/api/primitives.h [hunk 1]
```diff
@@ -52,21 +53,46 @@ z_string_t z_string_make(const char *value);
  */
 z_keyexpr_t z_keyexpr(const char *name);
 
+/**
+ * Constructs a :c:type:`z_keyexpr_t` departing from a string.
+ * It is a loaned key expression that aliases ``name``.
+ * Equal to z_zeyxepr: neither of them tests passed expression to correctness
```
Comment (cguimaraes) [inline include/zenoh-pico/api/primitives.h:59]: I do not get the point on this comment. Can you clarify?
Also, cite other functions using :c:func:\`z_zeyxepr\` block.

File: include/zenoh-pico/collections/string.h [hunk 2]
```diff
@@ -90,8 +90,8 @@ _z_string_t _z_string_from_bytes(const _z_bytes_t *bs);
  *   char **_val: A pointer to the array.
  */
 typedef struct {
-    size_t _len;
-    char **_val;
+    char **val;
+    size_t len;
 } _z_str_array_t;
```
Comment (cguimaraes) [inline include/zenoh-pico/collections/string.h:95]: Order types in structs from bigger size to smaller size.

**Review by cguimaraes — COMMENTED** [review]
LGTM, but some clarification asked

File: include/zenoh-pico/api/primitives.h [hunk 3]
```diff
@@ -43,6 +43,7 @@ z_string_t z_string_make(const char *value);
 /**
  * Constructs a :c:type:`z_keyexpr_t` departing from a string.
  * It is a loaned key expression that aliases ``name``.
+ * Equal to z_zeyxepr_unckecked: neither of them tests passed expression to correctness
```
[the remark this answers is not in the document]
Comment (milyin) [reply include/zenoh-pico/api/primitives.h:46]: In the zenoh-c the functions z_keyexpr differs from z_keyexpr_unchecked: the first one validates passed expression, senond one - don't. I wanted to point to the fact that the in the zenoh-pico validation is not performed in both functions. Will clarify the comment

File: include/zenoh-pico/api/primitives.h (hunk 1 above)
Comment (milyin) [reply include/zenoh-pico/api/primitives.h:59]: In the zenoh-c the functions z_keyexpr differs from z_keyexpr_unchecked: the first one validates passed expression, senond one - don't. I wanted to point to the fact that the in the zenoh-pico validation is not performed in both functions. Will clarify the comment

File: include/zenoh-pico/collections/bytes.h [hunk 4]
```diff
@@ -28,8 +28,8 @@
  *   uint8_t *val: A pointer to the bytes array.
  */
 typedef struct {
-    size_t len;
     const uint8_t *start;
+    size_t len;
```
[the remark this answers is not in the document]
Comment (milyin) [reply include/zenoh-pico/collections/bytes.h:32]: Field order should be the same as in zenoh-c, to make { .val = xxx, .len = yyy } initializers work. Ok, will undo this change and do the field order change in zenoh-c

File: include/zenoh-pico/collections/string.h (hunk 2 above)
Comment (milyin) [reply include/zenoh-pico/collections/string.h:95]: Field order should be the same as in zenoh-c, to make { .val = xxx, .len = yyy } initializers work. Ok, will undo this change and do the field order change in zenoh-c

File: include/zenoh-pico/protocol/core.h [hunk 5]
```diff
@@ -50,10 +50,14 @@ typedef struct {
  * A zenoh timestamp.
  */
 typedef struct {
-    _z_bytes_t _id;
-    uint64_t _time;
+    uint64_t time;
+    _z_bytes_t id;
```
[the remark this answers is not in the document]
Comment (milyin) [reply include/zenoh-pico/protocol/core.h:54]: Field order should be the same as in zenoh-c, to make { .val = xxx, .len = yyy } initializers work. Ok, will undo this change and do the field order change in zenoh-c

File: include/zenoh-pico/collections/string.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (milyin) [reply include/zenoh-pico/collections/string.h:95]: Yes, the zenoh-c have them public and zenoh-cpp uses that

File: include/zenoh-pico/protocol/core.h [hunk 6]
```diff
@@ -50,10 +50,14 @@ typedef struct {
  * A zenoh timestamp.
  */
 typedef struct {
```
[the remark this answers is not in the document]
Comment (milyin) [reply include/zenoh-pico/protocol/core.h:52]: Yes, the zenoh-c have them public and zenoh-cpp uses that

**Review by cguimaraes — APPROVED** [review]

**Pull request merged**